src/python/pants/backend/terraform/tffmt.py

```diff
@@ -0,0 +1,198 @@
+    )
+
+    downloaded_terraform, source_files = await MultiGet(
+        download_terraform_request, source_files_request
+    )
+
+    source_files_snapshot = (
+        source_files.snapshot
+        if setup_request.request.prior_formatter_result is None
+        else setup_request.request.prior_formatter_result
+    )
+
+    input_digest = await Get(
+        Digest,
+        MergeDigests((source_files_snapshot.digest, downloaded_terraform.digest)),
+    )
+
+    # `terraform fmt` operates on a directory-by-directory basis. First determine the directories in
+    # the snapshot. This does not use `source_files_snapshot.dirs` because that will be empty if the files
+    # are in a single directory.
+    directories = defaultdict(list)
+    for file in source_files.snapshot.files:
+        directory = os.path.dirname(file)
+        if directory == "":
+            directory = "."
+        directories[directory].append(file)
+
+    # Then create a process for each directory.
+    directory_to_process = {}
+    for directory, files_in_directory in directories.items():
+        args = [
+            "./terraform",
+            "fmt",
+            "-check" if setup_request.check_only else None,
+            directory,
+        ]
+        args = [arg for arg in args if arg]
+
+        process = Process(
+            argv=args,
+            input_digest=input_digest,
+            output_files=files_in_directory,
+            description=f"Run `terraform fmt` on {pluralize(len(files_in_directory), 'file')}.",
+            level=LogLevel.DEBUG,
+        )
+
+        directory_to_process[directory] = process
+
+    return Setup(
+        directory_to_process=directory_to_process, original_digest=source_files_snapshot.digest
+    )
+
+
+@rule(desc="Format with `terraform fmt`")
+async def tffmt_fmt(request: TffmtRequest, tffmt: TfFmtSubsystem) -> FmtResult:
+    if tffmt.options.skip:
+        return FmtResult.skip(formatter_name="tffmt")
+    setup = await Get(Setup, SetupRequest(request, check_only=False))
+    processes = [
+        (directory, Get(ProcessResult, Process, process))
+        for directory, process in setup.directory_to_process.items()
+    ]
+    results = await MultiGet(p[1] for p in processes)
+
+    def format(directory, output):
+        if len(output.strip()) == 0:
+            return ""
+
+        return textwrap.dedent(
+            """\
+        Output from `terraform fmt` on files in {directory}:
+        {output}
+
+        """
+        ).format(directory=directory, output=output.decode("utf-8"))
+
+    stdout_content = ""
+    stderr_content = ""
+    for i, result in enumerate(results):
+        process = processes[i][0]
+        stdout_content += format(process, result.stdout)
+        stderr_content += format(process, result.stderr)
+
+    # Merge all of the outputs into a single output.
+    output_digest = await Get(Digest, MergeDigests(r.output_digest for r in results))
```

Comment:
@stuhood: Your original comment here was for @Eric-Arellano: If the processes have been invoked independently, is it still necessary to merge into a single result? Or can this be "unbatched"?

Reply:
Yeah, it is necessary for now. This is interesting that there's a formatter that partitions. 

We could possibly add generic support for it, although I'm hesitant because partitioning is much riskier with `fmt` than `lint`. With `lint`, it's okay to run over the same file multiple times, e.g. Flake8 w/ Py2 and then Py3. But that is not safe with `fmt`, there must not be merge conflicts.